File: Gauss-Seidel Method/GaussSeidelWithConverge.py
import numpy as np


def gauss_seidel(A, B, max_iterations, tolerance=1e-10):
    n = len(B)
    X = np.zeros(n)
    prev_X = np.zeros(n)

    for k in range(max_iterations):
        prev_X[:] = X[:]  # Copy current solution to prev_X

        for i in range(n):
            sum_ = sum(A[i][j] * X[j] for j in range(n) if j != i)
            X[i] = (B[i] - sum_) / A[i][i]

        # Check for division by zero
        if any(X == 0):
            print("Error: Division by zero encountered.")
            return

        # Calculate error percentage
        errors = [abs((X[i] - prev_X[i]) / X[i]) * 100 if X[i]
                  != 0 else 0 for i in range(n)]

        # Check for convergence
        if max(errors) < tolerance:
            print("Solution (Iteration {}):".format(k + 1))
            for i in range(n):
                print("X[{}] = {:.6f}, Error: {:.6f}%".format(
                    i + 1, X[i], errors[i]))
            return

    print("The method did not converge within the specified number of iterations.")


# Input
n = int(input("Enter the number of equations: "))
A = [[0] * n for _ in range(n)]
B = [0] * n

print("Enter the coefficients of the equations (A matrix):")
for i in range(n):
    A[i] = list(map(float, input(
        "Equation {} coefficients (space-separated): ".format(i + 1)).split()))

print("Enter the constants (B matrix):")
for i in range(n):
    B[i] = float(input("Constant for equation {}: ".format(i + 1)))

max_iterations = int(input("Enter the number of iterations: "))

# Solve
gauss_seidel(A, B, max_iterations)

File: Gauss-Seidel Method/GaussSeidelWithConverge.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int is_diagonally_dominant(int x[][4], int n);

void seidel(int a[][4], float x[], int b[], int n);

float calculate_error(float previous[], float current[], int n);

int main()
{
  int n = 4; // Number of variables
  int a[4][4] = {{900, 7, 5, 8},
                 {11, 1700, 100, -20},
                 {17, -2, 600, -7},
                 {6, 7, 110, 1100}}; // Coefficient matrix
  int b[4] = {-70, 100, 60, 20};     // Constants matrix
  float x[4] = {0};                  // Initial solution

  printf("Initial Solution:");
  for (int i = 0; i < n; i++)
  {
    printf(" %.2f", x[i]);
  }
  printf("\n");

  if (is_diagonally_dominant(a, n))
  {
    float previous_x[4];
    for (int i = 0; i < n; i++)
    {
      previous_x[i] = x[i];
    }
    for (int i = 0; i < 300000; i++)
    {
      seidel(a, x, b, n);
      float error = calculate_error(previous_x, x, n);
      printf("Iteration %d:", i + 1);
      for (int j = 0; j < n; j++)
      {
        printf(" %.6f", x[j]);
      }
      printf(" Error: %.2f%%\n", error);
      for (int j = 0; j < n; j++)
      {
        previous_x[j] = x[j];
      }
    }
  }
  else
  {
    printf("The matrix is not diagonally dominant. The Gauss-Seidel method may not converge.\n");
  }

  return 0;
}

int is_diagonally_dominant(int x[][4], int n)
{
  for (int i = 0; i < n; i++)
  {
    int total = 0;
    for (int j = 0; j < n; j++)
    {
      if (i != j)
      {
        total += abs(x[i][j]);
      }
    }
    if (abs(x[i][i]) < total)
    {
      return 0;
    }
  }
  return 1;
}

void seidel(int a[][4], float x[], int b[], int n)
{
  float temp[4];
  for (int j = 0; j < n; j++)
  {
    float d = b[j];
    for (int i = 0; i < n; i++)
    {
      if (j != i)
      {
        d -= a[j][i] * x[i];
      }
    }
    temp[j] = d / a[j][j];
  }
  for (int i = 0; i < n; i++)
  {
    x[i] = temp[i];
  }
}

float calculate_error(float previous[], float current[], int n)
{
  float error = 0;
  for (int i = 0; i < n; i++)
  {
    float current_error = fabs((current[i] - previous[i]) / current[i]) * 100;
    if (current_error > error)
    {
      error = current_error;
    }
  }
  return error;
}